### ISSUE TITLE
The shader breaks on Iris beta + AMD.

### ISSUE THREAD

**Issue by NyaomiDEV** [Null-MC/Shrimple#4]
Trying to run RTX preset just for the sake of it fails on Iris 1.6.0 beta 5, using 1.19.4.
If that helps, I am using Windows, and an AMD card with driver version 23.3.2.

```
[Render thread/ERROR]: Failed to create shader rendering pipeline, disabling shaders!
net.coderbot.iris.gl.shader.ShaderCompileException: shadow_terrain_cutout.gsh: ERROR: 0:321: 'atomicOr' : atomic operations only can perform on buffer or shared storage. 
ERROR: 0:321: '' : compilation terminated 
ERROR: 2 compilation errors.  No code generated.
```

**Comment by null511** [comment]
I've seen this issue before, but only on AMD. It also only seems to happen with a specific (latest?) version of the drivers, but I don't have AMD myself to test with so I'm not entirely certain rolling back will fix.

**Comment by null511** [comment]
following up, I've been told AMD driver version `22.5.1` should be fine.

**Comment by NyaomiDEV** [comment]
AMD driver version 22.5.1 is ancient by today's standards, though. Bleeding edge and top tier cards users will either not be able to use that driver pack at all or take huge performance hits across the board; especially so after the release of driver package 23.2.2 which merged development efforts between RDNA2 cards (which now go super fast - the AMD FineWine joke still present) and RDNA3 cards.

I was informed that the bug is totally fixable, at least it was for some people. For example [the NobleRT developer fixed it in their shaderpack recently.](https://github.com/BelmuTM/NobleRT/commit/860c7149d697478bb6e31315d6b7d77eaf58a8c7)

As for me, I would gladly wait until this issue disappears either with time or with a contribution, as I can't rollback to that driver package version.

**Comment by null511** [comment]
> I was informed that the bug is totally fixable, at least it was for some people. For example [the NobleRT developer fixed it in their shaderpack recently.](https://github.com/BelmuTM/NobleRT/commit/860c7149d697478bb6e31315d6b7d77eaf58a8c7)

This is completely unrelated; Belmu is not using SSBO's or atomic operations at all, which is what is crashing for you. There are tons of different issues that can occur with AMD, he fixed something else specific to his usage. The _only_ report of the specific error you are running into was around using newer AMD drivers, it works fine on all other vendors.

As far as I'm aware, there isn't anything I can do to fix this on my end.

**Comment by tajemniktv** [comment]
Hiya, amd user here (23.3.1). Can confirm that it also doesn't work on my end. 
Many shaders have issues with amd cards, sadly... I'm guessing that won't help much or even at all, but here are some logs.
Full logs: https://mclo.gs/2K1odrR
Just the error: https://mclo.gs/9Ycs7Z4

Eh, screw amd XD

**Comment by DeadbushCarrot** [comment]
Yeah if shadow type is not set to default it just crashes. I got a rx 5500xt

**Comment by null511** [comment]
@DeadbushCarrot Are you using Optifine? And is this the last release [0.7] or the in-dev code download from github?

**Comment by DeadbushCarrot** [comment]
> @DeadbushCarrot Are you using Optifine? And is this the last release [0.7] or the in-dev code download from github?

Am using Iris 1.6.1 and 0.7 from modrinth

**Comment by null511** [comment]
Try using the latest in-dev version: https://github.com/Null-MC/Shrimple/archive/refs/heads/main.zip
most if not all those issues should be fixed. I'm in the middle of preparing a 0.8 release rn...

**Comment by DeadbushCarrot** [comment]
Nether darkness is fixed. transparent enchants are fixed. subsurface scattering is fixed. glowing rain is fixed. ok but the flat gray thing is still there if disappears when clouds are turned off. porosity is broken when set to Integrated. colored lighting is broken only when placed. there weird black squares below entity's. Block light is slightly better per-vertex doesn't crash but now it disable the shader until switched per-pixel crashes and so does ray-tracing.

**Comment by null511** [comment]
I've made some changes, and Shrimple should now work with the latest AMD drivers.
Basically AMD only broke `atomicOr` for SSBO's, but they still work fine on CU's, so I moved the voxel data there.

**Comment by DeadbushCarrot** [comment]
Porosity is fixed. the weird black squares below entity's are fixed. Block lighting is better again but it still doesn't work but now it won't crash your game just disables the shader. the glitch cloud is still there. colored lighting is the same as before

**Comment by null511** [comment]
This was using the latest github code download, and not release 0.8, right? The AMD changes have been pushed up, but not included into an official release yet.

**Comment by DeadbushCarrot** [comment]
Nope it was 0.8 i dont know how fast your pushing alphas out aka the GitHub version

**Comment by null511** [comment]
yeah my bad I should have specified that originally. I just posted the 0.8 release the other day, but found/made the AMD fix today.

**Comment by DeadbushCarrot** [comment]
VERTEX IS WORKING BUT BUGGY AS HELL

**Comment by DeadbushCarrot** [comment]
uhh ray tracing is wack
![2023-04-24_21 29 18](https://user-images.githubusercontent.com/109874700/234174758-a409f990-941b-4caa-85ee-11b0ba0bc8c1.png)

**Comment by null511** [comment]
Sorry about that, I indeed broke things in the latest commit making the buffers dynamically sized. I've fixed that now so the latest commit should be good.

**Comment by DeadbushCarrot** [comment]
We're getting there 
![2023-04-25_00 04 46](https://user-images.githubusercontent.com/109874700/234200448-fee81ec3-6564-4318-8781-4f4dbcf4dbaa.png)
![2023-04-25_00 07 05](https://user-images.githubusercontent.com/109874700/234200454-2623c788-78df-41f2-9577-b4b72cffe165.png)

**Comment by null511** [comment]
If you have a lot of lights, you may need to tweak the settings a bit, try:

Block Lighting > Advanced:
- Max Bin Lights = 128 (or more if needed)
- Bin Size = 4

Using a smaller bin-size allows more dense lighting. Increase the max count as needed, but more lights means less performance too.

**Comment by DeadbushCarrot** [comment]
also why does the torch's flicker it color really hard it gets close to green

**Comment by NyaomiDEV** [comment]
@DeadbushCarrot you should really make a new issue for your issues, as this issue outlines incompatibility with shaders on AMD cards running recent Windows drivers. As your setup works, you're more than likely on an older, supported driver version.

It'd be useful to clear up some confusion.

**Comment by NyaomiDEV** [comment]
wait are you really fixing the atomics?

**Comment by null511** [comment]
> also why is the torch flicker really hard it gets really close to green
weird, I'll check that out. You can also disable the brightness/hue change by turning off "flicker" in block light settings.

> wait are you really fixing the atomics?
Should already be fixed in the latest github commit (not release).

**Comment by DeadbushCarrot** [comment]
> > also why is the torch flicker really hard it gets really close to green
> 
> weird, I'll check that out. You can also disable the brightness/hue change by turning off "flicker" in block light settings.
> 
> > wait are you really fixing the atomics?
> 
> Should already be fixed in the latest github commit (not release).

No I ment in general you could make a option to turn on the flicker

**Comment by DeadbushCarrot** [comment]
> Tested it. I don't know on RDNA and other cards, but on my card it stutters a lot, in a very similar way to what I got with another shader, Rethinking Voxels. I am starting to think that either my card is horrendously damaged for Minecraft shader usage, or that AMD did multiple f-ups with their drivers.

Whats your gpu?

**Comment by DeadbushCarrot** [comment]
> > Whats your gpu?
> 
> RX 6900 XT

Am still in school but when I get home I tell what you should change in your drivers also is your drivers up to date?

**Comment by NyaomiDEV** [comment]
> Am still in school but when I get home I tell what you should change in your drivers also is your drivers up to date?

I think there's nothing wrong with my setup - and of course drivers are up to date. I made this very issue *because* I had drivers up to date, and I am still updating.

**Comment by null511** [comment]
There's several things that can be affecting performance, voxelizing and ray-tracing the world is also not a cheap technique in general. A good way to test your "peak performance" is by creating a flat world, and placing 1 to 3 lights and a few blocks for testing.

As for gameplay scenarios, you can try reducing some of the block-light bin settings:
- Max Bin Lights: Lowering this will improve performance, but allow less lights per bin/area.
- Bin Horizontal/Vertical Size: Lowering these will improve performance, but reduce the distance from camera that dynamic lighting is applied.

Also block-light VL is currently _very_ slow; best to avoid using it until I finish implementing the flood-fill LPV. It's off by default.

**Comment by DeadbushCarrot** [comment]
> 

was trying to help i was gunna say try turning off anit-lag also funny how my drivers aren't update to date

**Comment by null511** [comment]
Closing this now since AMD compatibility has been confirmed with the latest github main branch. This will be released soon as part of the 1.0 release of Shrimple.
